sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -1095,7 +1095,11 @@ private[hive] object HiveClientImpl extends Logging {
     table.bucketSpec match {
       case Some(bucketSpec) if !HiveExternalCatalog.isDatasourceTable(table) =>
         hiveTable.setNumBuckets(bucketSpec.numBuckets)
-        hiveTable.setBucketCols(bucketSpec.bucketColumnNames.toList.asJava)
```

Comment:
shall we just lower-case the bucket column names?

Reply:
The issue here is that this `toHiveTable()` is called 2 times during the test below.
First when the table is created. At that time the `table.schema` contains uppercase `B_C` and so does `table.bucketSpec` too (`B_C`). So simply lowercasing `bucketColumnNames` here for `setBucketCols()`  would throw a similar exception to the one in the description, but bucketspec would be lower and schema column would be uppercase.

Then 2nd time during the `collect()`, the Hive table is restored from metastore for a `listPartitionsByFilter()` Hive call. But this time `schema` contains lowercase `b_c` (columns are not case preserved) but `bucketSpec` contains uppercase `B_C` (bucket spec is case preserved for some reason) and `setBucketCols()` throws the exception in the description. I'm trying to fix this issue.

Reply:
AFAIK Spark stores table schema and bucket spec as table properties, and restores it back when reading the table from Hive metastore. So it is always case preserving in Spark, for both schema and bucket spec. Is there something wrong?

Reply:
That's true, but in `listPartitionsByFilter()` we call Hive's `getPartitionsByFilter()` with the raw table where we haven't restored schema/bucketspec from table properties: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L1303
And as far as I see besides `listPartitionsByFilter()`, some other methods of `HiveExternalCatalog` has the same issue...

Reply:
this seems like a bug. We already did `val catalogTable = restoreTableMetadata(rawTable)` and why don't we pass `catalogTable`?

Reply:
I guess we expected Hive to be consistent. If it is not case preseving with columns then why it is with bucket spec?

But there are other methods too, for example `renameTable` where we dont' trestore the table: [getRawTable]https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L539-L565

Reply:
I can't recall why we need to pass `rawTable` for some cases... Let's go with a different direction: Spark assumes that HMS always lower-case table/column names, and it's a bit tricky if bucket spec does not lower-case column names. Let's help Hive to fix it in https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L1114-L1118 by adding a `.map(_.toLowerCase)`

Reply:
Hmm, that's what I explained in https://github.com/apache/spark/pull/36027#discussion_r966199347 that simply lowercasing only bucketspec doesn't work. Because when the test creates the table and calls `toHiveTable()` first, it is called with uppercase column and bucketspec from the provided query text. If we lowecase bucketspec only then the failure will come during create table.

As Hive expects `setFields()` and `setBucketCols()` to use the same letter case we chould lowecase both or use my PR to sync bucketspec to columns.

Reply:
then let's lower case normal column names as well? Just don't trust the Hive metastore and normalize all names at Spark side.